src/components/tl/mlx5/tl_mlx5_pd.c

```diff
@@ -0,0 +1,283 @@
+    socklen_t          addrlen;
+
+    for (i = 0; i < group_size; i++) {
+        /* accept incoming connections */
+        connection[i].fd        = command_fd;
+        connection[i].pd_handle = pd_handle;
+        connection[i].sock      = accept(sock, NULL, 0);
+        if (connection[i].sock == -1) {
+            tl_error(lib,
+                     "failed to accept socket connection request %d,"
+                     " errno %d",
+                     i, errno);
+            goto listen_fail;
+        }
+        status = do_sendmsg(&connection[i]);
+        if (status != UCC_OK) {
+            tl_error(lib, "failed to send cmd_fd");
+            goto listen_fail;
+        }
+    }
+
+    addrlen = sizeof(addr);
+    getsockname(sock, (struct sockaddr *)&addr, &addrlen);
+
+listen_fail:
+    if (close(sock) == -1) {
+        tl_error(lib, "failed to close server socket errno %d", errno);
+        status = UCC_ERR_NO_MESSAGE;
+    }
+
+    if (remove(addr.sun_path) == -1) {
+        tl_error(lib, "socket file removal failed");
+        status = UCC_ERR_NO_MESSAGE;
+    }
+
+    return status;
+}
+
+ucc_status_t ucc_tl_mlx5_share_ctx_pd(ucc_tl_mlx5_context_t *ctx,
+                                      const char *           sock_path,
+                                      ucc_rank_t group_size, int is_ctx_owner,
+                                      int ctx_owner_sock)
+{
+    ucc_tl_mlx5_lib_t *lib =
+        ucc_derived_of(ctx->super.super.lib, ucc_tl_mlx5_lib_t);
+    int          ctx_fd;
+    uint32_t     pd_handle;
+    ucc_status_t status;
+
+    if (!is_ctx_owner) {
+        status = client_recv_data(&ctx_fd, &pd_handle, sock_path, lib);
+        if (UCC_OK != status) {
+            tl_error(lib, "failed to share ctx & pd from client side");
+            return status;
+        }
+        ctx->shared_ctx = ibv_import_device(ctx_fd);
+        if (!ctx->shared_ctx) {
+            tl_error(lib, "Import context failed");
+            return UCC_ERR_NO_MESSAGE;
+        }
+        ctx->shared_pd = ibv_import_pd(ctx->shared_ctx, pd_handle);
+        if (!ctx->shared_pd) {
+            tl_error(lib, "import PD failed");
+            if (ibv_close_device(ctx->shared_ctx)) {
+                tl_error(lib, "imported context close failed");
+            }
+            return UCC_ERR_NO_MESSAGE;
+        }
+        ctx->is_imported = 1;
+    } else {
+        ctx_fd    = ctx->shared_ctx->cmd_fd;
+        pd_handle = ctx->shared_pd->handle;
+        status = server_send_data(ctx_fd, pd_handle, group_size - 1,
+                                  ctx_owner_sock, lib);
+        if (UCC_OK != status) {
+            tl_error(lib, "failed to share ctx & pd from server side");
+            return status;
+        }
+    }
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_mlx5_remove_shared_ctx_pd(ucc_tl_mlx5_context_t *ctx)
+{
+    if (ctx->shared_pd) {
+        if (ctx->is_imported) {
+            ibv_unimport_pd(ctx->shared_pd);
```

Comment:
check status

Reply:
this function returns void (and does not set `errno`).
Do you have something else in mind ?